boards: add Pinnacle 100 DVK

rerickson1 opened the pull request:
Add Laird Connectivity Pinnacle 100 modem development kit.

This depends on https://github.com/zephyrproject-rtos/zephyr/pull/27676

carlescufi commented on boards/arm/pinnacle_100_dvk/pinnacle_100_dvk_defconfig:
```diff
@@ -0,0 +1,40 @@
+CONFIG_MODEM_HL7800=y
+
+# Networking
+CONFIG_NETWORKING=y
```
Are you sure you want networking enabled by default? You could instead leave it to the user

rerickson1 replied:
We feel the point of using our modem is to use it for networking, thus enabling it by default to make it easier for the user out of the gate.

jfischer-phytec-iot replied:
I will block this PR because of `CONFIG_NETWORKING=y` and `CONFIG_NET_IPV6=n`, which leads to unnecessary changes in tests and samples. This board is based on nRF52840 and therefore also supports Bluetooth LE, 802154, and IPv6. If nRF52840 does not have any RF functionality, then the items "ble, ieee802154" must be removed. 

And in general it is not necessary to enable CONFIG_NETWORKING on board level. Which networking sample does not do it anyway?

ioannisg reviewed (approved):
Looks fine to me.
Guess it is weird how this was passing CI with .yaml file having a wrong name.
Also, it is the first time I see slot1_partition being defined on a separate memory container :)

carlescufi commented:
@rerickson1 please take a look at the CI failures

rerickson1 commented:
> @rerickson1 please take a look at the CI failures

I believe all the CI errors relate to the HL7800 driver not being present yet.  Will rebase this once #27676 is approved and merged.

MaureenHelm commented on boards/arm/bl654_dvk/doc/bl654_dvk.rst:
```diff
@@ -1,188 +0,0 @@
-.. _bl654_dvk:
```
Did you mean to remove this file?

rerickson1 replied:
Rebase error, thanks for catching this.

MaureenHelm commented on boards/arm/pinnacle_100_dvk/doc/index.rst:
```diff
@@ -0,0 +1,251 @@
+Applications for the ``pinnacle_100_dvk`` board configuration can be
+built and flashed in the usual way (see :ref:`build_an_application`
+and :ref:`application_run` for more details); however, the standard
+debugging targets are not currently available.
```
This seems to be in conflict the debugging section below?

MaureenHelm commented on boards/arm/pinnacle_100_dvk/pinnacle_100_dvk.yaml:
```diff
@@ -1,19 +1,23 @@
-identifier: adafruit_feather_nrf52840
-name: Adafruit Feather nRF52840 Express
```
Did you mean to move this file? That will leave the Adafruit board without a yaml file

rerickson1 replied:
Rebase error

MaureenHelm commented on drivers/modem/hl7800.c:
```diff
@@ -4627,7 +4626,7 @@ static struct net_offload offload_funcs = {
 static inline uint8_t *hl7800_get_mac(struct device *dev)
 {
-	struct hl7800_iface_ctx *ctx = dev->driver_data;
+	struct hl7800_iface_ctx *ctx = dev->data;
```
The struct device changes should be in a separate commit from the newlibc changes. Both need a better commit message explanation.

rerickson1 committed:
drivers: modem: hl7800: remove VLA

Remove variable length array from function to comply with MISRA checks.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rerickson1 committed:
drivers: modem: hl7800: struct device data renamed

change struct device->driver_data to new name of struct device->data.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rerickson1 committed:
drivers: modem: hl7800: modem receiver IMEI

Modem receiver IMEI data field is only available
when CONFIG_MODEM_SIM_NUMBERS.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rerickson1 committed:
drivers: modem: hl7800: RTC only available with newlibc

RTC API is only available if NEWLIB_LIBC is available.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

MaureenHelm commented on samples/philosophers/sample.yaml:
```diff
@@ -30,6 +30,7 @@ tests:
   sample.kernel.philosopher.lifos:
     extra_args: "-DFORKS=LIFOS"
   sample.kernel.philosopher.preempt_only:
+    platform_exclude: pinnacle_100_dvk
```
why are you explicitly excluding this platform for so many samples? will you please explain in your commit message?

rerickson1 replied:
A reason for some of these is that I could not add a board kcofnig overlay and get it to work. I believe there is an issue.  I filed it here https://github.com/zephyrproject-rtos/zephyr/issues/27934.

MaureenHelm reviewed (commented):
thanks for splitting up the struct device and newlibc changes. just one more question

rerickson1 committed:
drivers: modem: hl7800: do not require DEVICE_POWER_MANAGEMENT

Remove requirement for DEVICE_POWER_MANAGEMENT,
the driver does not need it to operate.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

jfischer-phytec-iot commented on boards/arm/pinnacle_100_dvk/pinnacle_100_dvk.yaml:
```diff
@@ -0,0 +1,23 @@
+  - usb_cdc
+  - usb_device
+  - watchdog
+  - netif:openthread
```
If "Pinnacle 100 (HL7800 driver) does not support IPv6" then why is here "netif:openthread" present?

nashif commented on tests/kernel/sched/schedule_api/testcase.yaml:
```diff
@@ -10,6 +10,7 @@ tests:
       - CONFIG_TIMESLICING=n
     tags: kernel threads sched userspace
   kernel.scheduler.multiq:
+    platform_exclude: pinnacle_100_dvk
```
why is this being excluded? You need to add a comment about this.

nashif commented on tests/drivers/console/testcase.yaml:
```diff
@@ -14,4 +14,5 @@ tests:
     tags: drivers console
     arch_allow: arm
     filter: CONFIG_CPU_CORTEX_M and CONFIG_QEMU_TARGET
+    platform_exclude: pinnacle_100_dvk
```
why is this being excluded? You need to add a comment about this.

nashif reviewed (changes requested):
Why are you excluding all of those samples and tests, there is something not right here and no explanation of why those are being excluded.

lairdjm commented:
@jfischer-phytec-iot the issue was that the HL7800 driver doesn't support IPv6 and (in Zephyr 1.14 at least, hasn't been tested since) if any sample apps which had IPv6 sockets (e.g. http) in them were compiled and ran, they would instantly hardfault, the IPv6 Bluetooth radio was never tested. The board has networking enabled by default because it is a cellular modem module, it wouldn't make sense to use this particular board without networking.

rerickson1 committed:
drivers: modem: hl7800: add IPv6 support

Add support for dual mode IPv4/IPv6 stack.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rerickson1 committed:
drivers: modem: hl7800: fix dts binding

Fix hl7800 vendor prefix to match existing Sierra Wireless entry.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

rerickson1 committed:
boards: add Pinnacle 100 DVK

Add Laird Connectivity Pinnacle 100 modem.

Signed-off-by: Ryan Erickson <ryan.erickson@lairdconnect.com>

ioannisg commented:
@rerickson1 moving this to 2.5.0 release (you can add TSC-label to request handling this exceptionally and include it in 2.4.0)

rerickson1 commented:
> @rerickson1 moving this to 2.5.0 release (you can add TSC-label to request handling this exceptionally and include it in 2.4.0)

Moving back to 2.4.0 and requesting TSC approval. This PR was started before the merge window closed. This PR fixes a few issues that prevent the HL7800 driver from compiling. Also, this should be very low risk to get in 2.4.0. @nashif can I get you to re-review?

jfischer-phytec-iot commented:
> This PR fixes a few issues that prevent the HL7800 driver from compiling.

You could move the fixes to a new PR and not to take this PR as "hostage".

rerickson1 commented:
Approved for inclusion of v2.4.0 based on TSC vote on 9-16-2020.
